extended/src/main/java/io/kubernetes/client/extended/kubectl/KubectlRolloutHistory.java

```diff
@@ -154,8 +170,108 @@ private void deploymentViewHistory(V1Deployment deployment, AppsV1Api api) throw
+  }
+
+  private void parseHistory(List<V1ControllerRevision> historyList, PodTemplateParser parser)
+      throws ApiException {
+    Map<Long, V1ControllerRevision> historyInfo = new HashMap<>();
+    for (V1ControllerRevision history : historyList) {
+      historyInfo.put(history.getRevision(), history);
+    }
+
+    if (revision > 0) {
+      V1ControllerRevision history = historyInfo.get(revision);
+      if (history == null) {
+        throw new ApiException("unable to find the specified revision " + revision);
+      }
+      template = parser.parse(history);
+      return;
+    }
+
+    List<Long> revisions = new ArrayList<>(historyInfo.keySet());
+    revisions.sort(Long::compareTo);
+    for (Long revision : revisions) {
+      String changeCause =
+          historyInfo.get(revision).getMetadata().getAnnotations().get(CHANGE_CAUSE_ANNOTATION);
+      if (changeCause == null || changeCause.isEmpty()) {
+        changeCause = "<none>";
+      }
+      histories.add(new History(revision, changeCause));
+    }
+  }
+
+  // controlledHistories returns all ControllerRevisions in namespace that selected by selector and
+  // owned by accessor
+  private List<V1ControllerRevision> controlledHistory(
+      AppsV1Api api, KubernetesObject accessor, LabelSelector selector) throws ApiException {
+    List<V1ControllerRevision> result = new ArrayList<>();
+    V1ControllerRevisionList historyList =
+        api.listNamespacedControllerRevision(
+            namespace, null, null, null, null, selector.toString(), null, null, null, null, null);
+    for (V1ControllerRevision history : historyList.getItems()) {
+      if (isControlledBy(history, accessor)) {
+        result.add(history);
+      }
+    }
+    return result;
+  }
+
+  private boolean isControlledBy(KubernetesObject obj, KubernetesObject owner) {
+    Optional<V1OwnerReference> ref =
+        obj.getMetadata().getOwnerReferences().stream()
+            .filter(r -> r.getController() != null && r.getController())
+            .findAny();
+    return ref.map(
+            v1OwnerReference -> v1OwnerReference.getUid().equals(owner.getMetadata().getUid()))
+        .orElse(false);
+  }
+
   // getChangeCause returns the change-cause annotation of the input object
   private static String getChangeCause(V1ObjectMeta meta) {
     return meta.getAnnotations().get(CHANGE_CAUSE_ANNOTATION);
   }
+
+  // now we just support directive in merge
+  private static String applyDirectivePatch(String origin, String patch) {
+    // todo StrategicPatch?
+    Map<String, Object> patchMap = new JSON().deserialize(patch, Map.class);
+    Map<String, Object> specMap = (Map<String, Object>) patchMap.get("spec");
+    Map<String, Object> templateMap = (Map<String, Object>) specMap.get("template");
+    final String directiveMarker = "$patch";
+    Object directive = templateMap.get(directiveMarker);
+    if (directive == null) {
+      throw new IllegalArgumentException("not a directive patch");
```

Comment:
i dont think the SMP directives are mandatory in the patch body, and as SSA(Server) goes GA, we should switch all the usage of SMP to SSA to get rid of these magic markers. can you add a TODO comment over here saying that we prefer to land onto SSA patches in a follow-up?